test/end-to-end-tests/src/rest/session.ts

```diff
@@ -0,0 +1,137 @@
+limitations under the License.
+*/
+
+import request = require('request-promise-native');
+import { Logger } from '../logger';
+import { RestRoom } from './room';
+import { approveConsent } from './consent';
+import { Credentials } from "./creator";
+
+interface RoomOptions {
+    invite: string;
+    public: boolean;
+    topic: string;
+    dm: boolean;
+}
+
+export class RestSession {
+    private _displayName: string = null;
+    private readonly rooms: Record<string, RestRoom> = {};
+    readonly log: Logger;
+
+    constructor(private readonly credentials: Credentials) {
+        this.log = new Logger(credentials.userId);
+    }
+
+    userId(): string {
+        return this.credentials.userId;
+    }
+
+    userName(): string {
+        return this.credentials.userId.split(":")[0].substr(1);
+    }
+
+    displayName(): string {
+        return this._displayName;
+    }
+
+    async setDisplayName(displayName: string): Promise<void> {
+        this.log.step(`sets their display name to ${displayName}`);
+        this._displayName = displayName;
+        await this.put(`/profile/${this.credentials.userId}/displayname`, {
+            displayname: displayName,
+        });
+        this.log.done();
+    }
+
+    async join(roomIdOrAlias: string): Promise<RestRoom> {
+        this.log.step(`joins ${roomIdOrAlias}`);
+        const roomId = (await this.post(`/join/${encodeURIComponent(roomIdOrAlias)}`)).room_id;
+        this.log.done();
+        const room = new RestRoom(this, roomId, this.log);
+        this.rooms[roomId] = room;
+        this.rooms[roomIdOrAlias] = room;
+        return room;
+    }
+
+    room(roomIdOrAlias: string): RestRoom {
+        if (this.rooms.hasOwnProperty(roomIdOrAlias)) {
+            return this.rooms[roomIdOrAlias];
+        } else {
+            throw new Error(`${this.credentials.userId} is not in ${roomIdOrAlias}`);
+        }
+    }
+
+    async createRoom(name: string, options: RoomOptions): Promise<RestRoom> {
+        this.log.step(`creates room ${name}`);
+        const body = {
+            name,
+        };
+        if (options.invite) {
+            body['invite'] = options.invite;
+        }
+        if (options.public) {
+            body['visibility'] = "public";
+        } else {
+            body['visibility'] = "private";
+        }
+        if (options.dm) {
+            body['is_direct'] = true;
+        }
+        if (options.topic) {
+            body['topic'] = options.topic;
+        }
+
+        const roomId = (await this.post(`/createRoom`, body)).room_id;
+        this.log.done();
+        return new RestRoom(this, roomId, this.log);
+    }
+
+    post(csApiPath: string, body?: any): Promise<any> {
+        return this.request("POST", csApiPath, body);
+    }
+
+    put(csApiPath: string, body?: any): Promise<any> {
+        return this.request("PUT", csApiPath, body);
+    }
+
+    async request(method: string, csApiPath: string, body?: any): Promise<any> {
```

Comment:
Can we do better here, maybe `Record<string, any>` at the very least

Reply:
well its a JSON response, so it could be a scalar or `[]` .. effectively, `any`

Reply:
I don't think any Matrix APIs return `[]` or `null` though

Reply:
the types should reflect the deterministic reality .. some matrix API could return that in future, or the server could choose to send back plaintext due to a config error, etc etc

until we actually validate the value, its an `any`